src/Build/BackEnd/Components/SdkResolution/MainNodeSdkResolverService.cs

```diff
@@ -0,0 +1,240 @@
+        /// <param name="projectPath">The full path to the project that referenced the SDK.</param>
+        /// <returns>An <see cref="SdkResult"/> containing information about the SDK if one was resolved, otherwise <code>null</code>.</returns>
+        private SdkResult GetSdkResult(int submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, string solutionPath, string projectPath)
+        {
+            ErrorUtilities.VerifyThrowInternalNull(sdk, nameof(sdk));
+            ErrorUtilities.VerifyThrowInternalNull(loggingContext, nameof(loggingContext));
+            ErrorUtilities.VerifyThrowInternalNull(sdkReferenceLocation, nameof(sdkReferenceLocation));
+
+            // Get the dictionary for the specified submission if one is already added otherwise create a new dictionary for the submission.
+            ConcurrentDictionary<string, SdkResult> cached = _cache.GetOrAdd(submissionId, new ConcurrentDictionary<string, SdkResult>(MSBuildNameIgnoreCaseComparer.Default));
+
+            // Get a cached result if available, otherwise resolve the SDK with the SdkResolverService.Instance
+            SdkResult result = cached.GetOrAdd(
+                sdk.Name,
+                key =>
+                {
+                    SdkResult sdkResult = SdkResolverService.Instance.GetSdkResult(sdk, loggingContext, sdkReferenceLocation, solutionPath, projectPath);
+
+                    // Associate the element location of the resolved SDK reference
+                    sdkResult.ElementLocation = sdkReferenceLocation;
+
+                    return sdkResult;
+                });
+
+            if (!SdkResolverService.IsReferenceSameVersion(sdk, result.Version))
+            {
+                // MSB4240: Multiple versions of the same SDK "{0}" cannot be specified. The SDK version already specified at "{1}" will be used and the version will be "{2}" ignored.
+                loggingContext.LogWarning(null, new BuildEventFileInfo(sdkReferenceLocation), "ReferencingMultipleVersionsOfTheSameSdk", sdk.Name, result.ElementLocation, sdk.Version);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Handles a request from a remote node.
+        /// </summary>
+        /// <param name="node">The ID of the remote node.</param>
+        /// <param name="request">The <see cref="SdkResolverRequest"/> containing information about the SDK to resolve.</param>
+        /// <remarks>This method must not directly handle requests because it would block requests from other nodes.  Instead, it simply
+        /// adds requests to a queue which are processed by a background thread.</remarks>
+        private void HandleRequest(int node, SdkResolverRequest request)
+        {
+            // Start the background thread which will process queued requests if it has not already been started.
+            lock (_lockObject)
```

Comment:
>lock [](start = 12, length = 4)

Duplicate the `if (_requestHandler == null)` to avoid locking

Reply:
Do you think it would be better to use a `Lazy<T>` instead?  I was worried about overhead but its only a single object per process.  Or I can just check `_requestHandler` twice...